library/core/src/intrinsics.rs

```diff
@@ -1147,7 +1147,8 @@ extern "rust-intrinsic" {
     /// may lead to unexpected and unstable compilation results. This makes `transmute` **incredibly
     /// unsafe**. `transmute` should be the absolute last resort.
     ///
-    /// Transmuting pointers to integers in a `const` context is [undefined behavior][ub].
+    /// Transmuting pointers to integers in a `const` context is [undefined behavior][ub]
+    /// (unless the pointer was created from an integer with [`invalid`][crate::ptr::invalid] or this function).
```

Comment:
I think this should talk about the general *idea* of transmuting, rather than restricting it to specifically "this function".  Notably, as of https://github.com/rust-lang/rust/pull/113344/files#diff-4a1a6c5ac9bcf4c877565c6aa0f6d649a84907b9778c9b7cabebbc314acb8d06L9-R17 the slice iterators no longer use *literally* `transmute` to go from a pointer to an integer, so the goal mentioned in the OP wouldn't be met with a narrow reading of the proposed text.

Here's my attempt at a rephrasing that I hope captures similar intent:
```suggestion
    /// Transmuting pointers *to* integers in a `const` context is [undefined behavior][ub],
    /// unless the pointer was originally created by transmuting *from* an integer.
    /// (That includes this function specifically and helpers like [`invalid`][crate::ptr::invalid],
    /// but also semantically-equivalent conversions such as punning through `repr(C)` union fields.)
```

Reply:
> conversions such as punning through `repr(C)` union fields.
Does the Union need to be `repr(C)`, of so why?

Reply:
And would it be better to actually use a Union in the slice iterator?

Reply:
`union Foo { ptr: *const i32, int: usize }` doesn't guarantee that the fields are at the same offset in `repr(Rust)`, only in `repr(C)`, so I included the caveat here.

If you have a non-`C` union where you've checked that the fields are at the same offset, then it'd also be defined behaviour.  As would "transmuting" via `addr_of!(my_ptr).cast::<usize>().read()`, and many other such things.

Reply:
Ahh ok, Thank you, I'll use your Text then.

Reply:
> And would it be better to actually use a Union in the slice iterator?

I'd love to, but see the conversation starting with https://github.com/rust-lang/unsafe-code-guidelines/issues/73#issuecomment-1629529101 for why I don't.

Reply:
Should we state somewhere more prominently in the `transmute` docs that transmuting via this method is equivalent do doing it via union field punning or pointer type punning?

Reply:
Should the docs say something like:
This function is semantically equivalent to:
```rust
use core::mem::ManuallyDrop;
unsafe fn transmute<Src, Dst>(src: Src) -> Dst {
    #[repr(C)]
    union Conv<Src, Dst> {
        src: ManuallyDrop<Src>,
        dst: ManuallyDrop<Dst>,
    }

    let tmp = Conv::<Src, Dst> { src: ManuallyDrop::new(src) };
    ManuallyDrop::into_inner(tmp.dst)
}
```
or converting between types via pointer casts.

Reply:
Yeah, something like that. That should probably be a separate PR though since FCP here was only about the original question.

Reply:
Sure, though if we do a PR that says those are equivalent we should probably stay on the original text for this PR (or change to that in the other PR).
I could do that follow up PR for this if you want.